### PR TITLE
[15.0][ADD] queue_job_cron_jobrunner

### PR DESCRIPTION
**Pull request by ivantodorovich**
This module implements a simple ``queue.job`` runner using ``ir.cron`` triggers.

It's meant to be used on environments where the regular job runner can't be run, like on Odoo.sh.

Unlike the regular job runner, where jobs are dispatched to the HttpWorkers, jobs are processed on the CronWorker threads by the job runner crons. This is a design decision because:

* Odoo.sh puts HttpWorkers to sleep when there's no network activity
* HttpWorkers are meant for traffic. Users shouldn't pay the price of background tasks.

For now, it only implements the most basic features of the ``queue_job`` runner, notably no channel capacity nor priorities.
Please check the ROADMAP for further details.

----

For the easiest case, no configuration is required besides installing the module.

To avoid CronWorker CPU timeout from abruptly stopping the job processing cron, it's recommended to launch Odoo with ``--limit-time-real-cron=0``, to disable the CronWorker timeout altogether. (In Odoo.sh, this is done by default)

Parallel execution of jobs can be achieved by leveraging multiple ``ir.cron`` records:

* Make sure you have enough CronWorkers available (Odoo CLI ``--max-cron-threads``)
* Duplicate the ``queue_job_cron`` cron record as many times as needed, until you have as much records as cron workers.

----

Related to:
- https://github.com/OCA/queue/issues/371 (@thomaspaulb this pretty much implements your suggestion)
- https://github.com/OCA/queue/issues/2

ping @simahawk @guewen

### CODE REVIEW

File: queue_job_cron_jobrunner/models/queue_job.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+        Odoo will prevent concurrent cron jobs from running.
+        So, to eventually support parallel execution, we'd need to have (at least) the
+        same number of ir.crons records as cron workers.
+
+        All crons should be triggered at the same time.
+        """
+        crons = (
+            self.env["ir.cron"]
+            .sudo()
+            .search(
```
Comment (simahawk) [inline queue_job_cron_jobrunner/models/queue_job.py]: maybe we need a special flag on ir.cron

File: queue_job_cron_jobrunner/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
+    "name": "Queue Job Cron Jobrunner",
+    "summary": "Run jobs without a dedicated JobRunner",
+    "version": "15.0.1.0.0",
```
Comment (simahawk) [inline queue_job_cron_jobrunner/__manifest__.py:4]:
```suggestion
    "version": "15.0.1.0.0",
    "development_status": "Alpha",
```

File: queue_job_cron_jobrunner/models/queue_job.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+            """
+            SELECT *
+            FROM queue_job
+            WHERE state = 'pending'
+            AND (eta IS NULL OR eta <= (now() AT TIME ZONE 'UTC'))
+            ORDER BY date_created DESC
+            LIMIT 1 FOR NO KEY UPDATE SKIP LOCKED
+            """
+        )
+        row = self.env.cr.dictfetchone()
```
Comment (simahawk) [inline queue_job_cron_jobrunner/models/queue_job.py]:
```suggestion
        row = self.env.cr.fetchone()
```

File: queue_job_cron_jobrunner/models/queue_job.py [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+        except RetryableJobError as err:
+            # delay the job later, requeue
+            job.postpone(result=str(err), seconds=5)
+            job.set_pending(reset_retry=False)
+            job.store()
+            _logger.debug("%s postponed", job)
+
+        except (FailedJobError, Exception):
+            buff = StringIO()
```
Comment (simahawk) [inline queue_job_cron_jobrunner/models/queue_job.py]: close the buffer at the end

File: queue_job_cron_jobrunner/models/queue_job.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+        """Acquire the next job to be run.
+
+        :returns: queue.job record (locked for update)
+        """
+        # TODO: This method should respect channel priority and capacity,
+        #       rather than just fetching them by creation date.
+        self.flush()
+        self.env.cr.execute(
+            """
+            SELECT *
```
Comment (simahawk) [inline queue_job_cron_jobrunner/models/queue_job.py]:
```suggestion
            SELECT id
```

File: queue_job_cron_jobrunner/models/queue_job.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+            FROM queue_job
+            WHERE state = 'pending'
+            AND (eta IS NULL OR eta <= (now() AT TIME ZONE 'UTC'))
+            ORDER BY date_created DESC
+            LIMIT 1 FOR NO KEY UPDATE SKIP LOCKED
+            """
+        )
+        row = self.env.cr.dictfetchone()
+        return self.browse(row and row["id"])
```
Comment (simahawk) [inline queue_job_cron_jobrunner/models/queue_job.py]:
```suggestion
        return self.browse(row and row[0])
```

**Review by simahawk — APPROVED** [review]
LG and FTR we are testing it on odoo.sh these days ;)

**Review by guewen — APPROVED** [review]
Thank you!

**Comment by simahawk** [comment]
@ivantodorovich wanna squash or we merge like this?

**Comment by ivantodorovich** [comment]
@simahawk done ;)

**Comment by StefanRijnhart** [comment]
For the record, this allows you to run jobs on the 'hybrid model of multiple threaded workers' of Odoo.sh as mentioned in https://github.com/OCA/queue/issues/169, right?

**Comment by ivantodorovich** [comment]
That's right, by getting rid of the job runner. Instead it uses async `ir.cron`s to process jobs

**Comment by chrisb-c01** [comment]
I've (tried to) create a backport from 15.0 to 14.0. As known, 14.0 does not feature the `ir.cron.trigger` model that 15.0 introduced. Instead I'm just setting the `cron.nextcall` attribute to the current datetime (link to [code](https://github.com/OCA/queue/pull/479/commits/573cb3478dc32ff1658ba642cf70711fab97cbfc#diff-e0bfcb5a4868e4856304442bdf2a6372bbd3668e4101150804e280fbe8df5dabR149)). 

@ivantodorovich would be great if you could have a look at the [PR](https://github.com/OCA/queue/pull/479). Thanks for the initial commit.

**Comment by guewen** [comment]
Hi @amigrave, this is the module we spoke about in Brussels :)

**Pull request merged**